From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: SpreadZhao <spreadzhao@outlook.com>
Date: Thu, 29 Oct 2026 12:18:32 +0800
Subject: [PATCH] bar/scene-cleanup: destroy scene output before monitor
 teardown

Destroy m->scene_output before tearing down layer surfaces and bar buffers
so wlroots does not emit surface-enter events for an output being destroyed
(wlroots issue #4096).
---
 dwl.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/dwl.c b/dwl.c
index 12db97f..3690782 100644
--- a/dwl.c
+++ b/dwl.c
@@ -984,6 +984,12 @@ cleanupmon(struct wl_listener *listener, void *data)
 	LayerSurface *l, *tmp;
 	size_t i;
 
+	/* Stop surface-enter events for this output before changing the scene.
+	 * New listeners would miss the output destroy signal already in progress.
+	 * See https://gitlab.freedesktop.org/wlroots/wlroots/-/issues/4096 */
+	wlr_scene_output_destroy(m->scene_output);
+	m->scene_output = NULL;
+
 	/* m->layers[i] are intentionally not unlinked */
 	for (i = 0; i < LENGTH(m->layers); i++) {
 		wl_list_for_each_safe(l, tmp, &m->layers[i], link)
@@ -1004,7 +1010,6 @@ cleanupmon(struct wl_listener *listener, void *data)
 		destroylocksurface(&m->destroy_lock_surface, NULL);
 	m->wlr_output->data = NULL;
 	wlr_output_layout_remove(output_layout, m->wlr_output);
-	wlr_scene_output_destroy(m->scene_output);
 
 	closemon(m);
 	wlr_scene_node_destroy(&m->fullscreen_bg->node);
